compiler/codegen/cg-expr.cpp

```diff
@@ -5412,7 +5413,8 @@ static void codegenPutGet(CallExpr* call, GenRet &ret) {
       }
 
       // c_ptr/ddata are already addresses, so dereference one level.
-      if (dt->hasFlag(FLAG_DATA_CLASS)) {
+      // TODO: Understand why the deref is OK for some types but not for c_ptrConst(c_char)?
+      if (dt->hasFlag(FLAG_DATA_CLASS) && !isCPtrConstChar(dt->type)) {
```

Comment:
Hmm, this does not seem right. Maybe we should look at the problem you were facing together.

Reply:
I've created https://github.com/chapel-lang/chapel/pull/22884 to try to resolve this in a way that I find more satisfying. I did check that if I merge 22884 with your branch here, and then I fully remove the blocks that 22884 adds warnings to, then the reproducer you provided me compiles and runs successfully.

Reply:
rebased, updated the call sites in `DefaultRectangular.chpl` and removed the mentioned blocks in PR 22884